core/network/src/lib.rs

```diff
@@ -17,11 +17,148 @@
+//! - We periodically open ephemeral substreams for Kademlia random walk queries. Each Kademlia
+//! RPC query is done in a new separate substream. This uses the
+//! [standard libp2p Kademlia protocol](https://github.com/libp2p/specs/pull/108).
+//! - We optionally keep a substream alive for all Substrate-based communications. The name of the
+//! protocol negotiated is based on the *protocol ID* passed as part of the network configuration.
+//! This protocol ID should be unique for each chain and prevents nodes from different chains from
+//! connecting to each other. More information below.
+//!
+//! ## The Substrate substream
+//!
+//! Substrate uses a component named the **peerset manager (PSM)**. Through the discovery
+//! mechanism, the PSM is aware of the nodes that are part of the network and decides which nodes
+//! we should perform Substrate-based communications with. For these nodes, we open a connection
+//! if necessary and open a unique substream for Substrate-based communications. If the PSM decides
+//! that we should disconnect a node, then that substream is closed.
+//!
+//! For more information about the PSM, see the *substrate-peerset* crate.
+//!
+//! Note that at the moment there is no mechanism in place to solve the issues that arise where the
+//! two sides of a connection open the unique substream simultaneously. In order to not run into
+//! issues, only the dialer of a connection is allowed to open the unique substream. When the
+//! substream is closed, the entire connection is closed as well. This is a bug, and should be
+//! fixed by improving the protocol.
+//!
+//! Within the unique Substrate substream, messages encoded using *parity-codec* are exchanged.
+//! The detail of theses messages is not totally in place, but they can be found in the
+//! `message.rs` file.
+//!
+//! Once the substream is open, the first step is an exchange of a *status* message from both
+//! sides, containing information such as the chain root hash, head of chain, and so on.
+//!
+//! Communications within this substream include:
+//!
+//! - Syncing. Blocks are announced and requested to other nodes.
+//! - Light-client requests. When a light client requires information, a random node we have a
+//! substream open with is chosen, and the information is requested from it.
+//! - Gossiping. Used for example by grandpa.
+//! - Network specialization. The network protocol can be specialized through a template parameter
+//! of the network service. This specialization is free to send and receive messages with the
+//! remote. This is meant to be used by the chain that is being built on top of Substrate
+//! (eg. Polkadot).
+//!
+//! It is intended that in the future each of these components gets more isolated, so that they
+//! are free to open and close their own substreams, and so that syncing and light client requests
+//! are able to communicate with nodes outside of the range of the PSM.
```

Comment:
These things still need PSM even if they're on different substreams right? We need to punish someone not responding correctly to block requests or be able to ban a light client excessively requesting data (e.g. DoS'ing).

Reply:
Not responding correctly to requests is a protocol error that's punished by a disconnect.
As for DoSing, that's a good point.